// ae/project.js
var helpStr = 'project:\n'
+'getActiveComp()	如果当前项为合成，返回合成，否则返回false\n'
+'getSelectedLayers()	获取选中的图层，返回图层数组，否则返回false\n'
+'findLayer(name[, comp])	参数: name 图层名称，comp 合成	根据名称找到对应图层，否则返回false\n'
+'getItemById(id)	参数: id 数字id		通过item的id获取对应item，否则返回false\n'
+'lookItemInProjectPanel(item)	参数: item 合成		在窗口中找到并选中项目\n'
+'looklayerInComp(layer)	参数: layer 图层	显示合成并选中图层\n'
+'allowAccessFile()	保证ae允许写入文件并访问网络，否则返回false\n';

module.exports = {
	help : helpStr,
	getActiveComp: function () {//如果当前项为合成，返回合成，否则返回false
		var thisComp = app.project.activeItem;
		if(thisComp instanceof CompItem){
			return thisComp;
		}
		return null;
	},
	getSelectedLayers: function () {//获取选中的图层，返回图层数组，否则返回false
		var thisComp = app.project.activeItem;
		if(!(thisComp instanceof CompItem) || thisComp.selectedLayers.length == 0){
			return false;
		}
		return thisComp.selectedLayers;
	},
	findLayer : function(name, comp) {//根据名称找到对应图层，否则返回false
		//找到指定名称的图层
		if(!comp) var comp = app.project.activeItem;
		if(!(comp instanceof CompItem)) return false;
		
		var l = comp.layers;
		if(l.length == 0) return false;
		
		for(var i = 1; i <= l.length; i++) {
			if(l[i].name == name) return l[i];
		}
		return false;
	},
	getItemById: function (id) {//通过item的id获取对应item，否则返回false (id)
		for(var i = 1; i <= app.project.numItems; i++) {
			if(app.project.item(i).id == parseInt(id)) {
				return app.project.item(i);
			}
		}
	
		return false;
	},
	getFolderItemByName: function (name, parent) {//通过文件夹item的name获取对应文件夹item，否则返回false (id)
		var items = parent ? parent.items : app.project.items;
		for(var i = 1; i <= items.length; i++) {
			if(items[i] instanceof FolderItem) {
				if(items[i].name == name) {
					return items[i];
				}else {
					var t = arguments.callee(name, items[i]);
					if(t) return t;
				}
			}
		}
	
		return false;
	},
	getCompItemByName: function (name, parent) {//通过文件夹item的name获取对应文件夹item，否则返回false (id)
		var items = parent ? parent.items : app.project.items;
		for(var i = 1; i <= items.length; i++) {
			if(items[i] instanceof CompItem && items[i].name == name) {
				return items[i];
			}else if(items[i] instanceof FolderItem) {
				var t = arguments.callee(name, items[i]);
				if(t) return t;
			}
		}
	
		return false;
	},
	getFirstSelectedItem: function () {//通过文件夹item的name获取对应文件夹item，否则返回false (id)
		var item = app.project.selection;
		if(item.length > 0) return item[0];
		else return null;
	},
	lookItemInProjectPanel: function (item) {//在窗口中找到并选中项目 (项目)
		for(var i = 1; i <= app.project.numItems; i++) {
			if(app.project.item(i).selected) {
				app.project.item(i).selected = 0;
			}
		}
		item.selected = 1;
	},
	looklayerInComp: function (layer) {//在显示合成并选中图层 (图层)
		var myComp = layer.containingComp;
		if(parseInt(app.version) > 9) {//因为cs4不支持openInViewer()函数
			myComp.openInViewer();
		}
		var sl = myComp.selectedLayers;
		for(var i = 0; i < sl.length; i++) {
			sl[i].selected = 0;
		}
		layer.selected = 1;
	}
};
